Remove return in case of function pointer

xbauch opened the pull request:
Motivation: we want `remove_returns` to work even if not preceded by `remove_function_pointers`. For example:

```
int (* const fptbl2[])(void) = {g1, g2};

int func(int id1)
{
  t = fptbl2[id1];  // for this instruction the set is {"g1", "g2"}
  return t;
}
```
This PR exposes the function-pointer-target-collecting part of `remove_function_pointers` to be used in other analyses and the does so in `remove_returns`.

In `remove_returns` use the `possible_fp_targets_map` to be able to remove returns for calls to function pointers. `CALL x=*fp()` should become `CALL *fp(); ASSIGN x = (fp == &f00) ? return_value_f00 : (fp == &f01) ? ...`.

The first 3 commits are refactoring/preparation. The fourth introduces the interface. The fifth implements the new `do_function_call` for function pointers.

Then we refactor the `remove_function_pointers` to extract the collect-candidate-functions part and separate the modify-function-call part.

Finally -- now that the file is open -- we also refactor the internal `remove_function_pointers` without any functional change.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Extract do_function_call loop into separate method

No functional change.

Commit:
Refactor do_function_calls

Move the checks for is_function_call and does-return-anything outside; and
pre-compute the is_stub. do_function_call now always does something.

Commit:
Split do_function_call into two for stub/non-stub

Also extract the common behaviour to do_function_calls.

NlightNFotis reviewed (approved):
Looks good to me overall

Commit:
Set up the interface for fp_target map

The type and the function to return the map.

Commit:
Add do_function_call for function pointers

In case the function id is dereference we grab the set of possible targets (via
the possible_fp_targets_map) and pass it to the new do_function_call method.

There we build a cond_expr by iterating over the possible targets and adding a
case: (f==target ? target_return_value) to it. This cond_expr then becomes the
right-hand-side of the new assignment. We also have to dead all the return
values that were created.

Also building remove_function_pointerst requires messaget

And now so does remove_returnst, hence extending the call sites as well.

Commit:
Extract the collection of candidate functions

From remove_function_pointers to a separate function split into three:
1. get_funtion_pointer_targets .. the wrapper (for model/program)
2. refine_call_type .. for incomplete function types
3. try_remove_const_fp .. call remove_const_function_pointer

Then implement a wrapper that returns a map: function-name -> list-of-candidates

Commit:
Re-implement remove_function_pointers

Using the new candidate-collecting function.

Commit:
Refactor remove_function_pointers

Introducing two new member function:
1. build_new_code .. wraps the collection of function calls and gotos
2. remove_function_pointer_log .. wraps the logging and statistics

Setting the location now takes place inside the loop over candidate functions:
there is no second loop over newly introduced gotos.
